### ISSUE TITLE
haproxy router monitoring

### ISSUE THREAD

**Issue by kincl** [openshift/cluster-monitoring-operator#229]
The default HAProxy OpenShift router exposes prometheus metrics that should be scraped by the prometheus cluster monitoring.

**Comment by sichvoge** [comment]
The cluster-monitoring-operator is configured to watch all `openshift-*` namespaces for additional configurations such as ServiceMonitor CRDs. That way, we allow OpenShift components (e.g. HAproxy) to dynamically add their endpoints to the cluster-level Prometheus without us having to change anything inside the cluster-monitoring-operator necessary. 

Now, it's up to each OpenShift component team to actually deploy that configuration. Nothing we can do. That said, I am in touch with the HAproxy team to actually do exact that. :) Stay tuned.

Nevertheless, since it's not a concern for this project, we should close this issue.

**Comment by kincl** [comment]
Fair enough @sichvoge thanks for the detailed response :)

**Comment by alekonko** [comment]
any news?

**Comment by sichvoge** [comment]
This should work in the newest OpenShift 4.1 version.

**Comment by alekonko** [comment]
there is a plan to backport the isp-monitoring feature in 3.xx? Actually isn't clear the update path of cluster to a newest version.

actually in ocp 3.11 scraping haproxy stats from my Prometheus instance. sometime the exporter (included in haproxy 3.11) in production environment report us strange metric value ( sometime for few minutes 5xx and 2xx metrics rate value go to 100x  from example from 50-70 to 1000-5000). could be a problem of the exporter or a real problem of a haproxy instances?

**Comment by sichvoge** [comment]
I am working with the HAproxy team to understand the possibility of providing a supportable option for 3.11, but don't get your hopes to high up just yet :) It's difficult and backporting new features is difficult.

Re your question about HAproxy, better to contact the HAproxy team. I can't help you with that.

**Comment by alekonko** [comment]
thanks anyway

**Comment by alibo** [comment]
> actually in ocp 3.11 scraping haproxy stats from my Prometheus instance. sometime the exporter (included in haproxy 3.11) in production environment report us strange metric value ( sometime for few minutes 5xx and 2xx metrics rate value go to 100x from example from 50-70 to 1000-5000). could be a problem of the exporter or a real problem of a haproxy instances?

@alekonko it's because multiple haproxy processes are running with different stats (it happens during reloading haproxy and there are long running connections)! so sometimes prometheus scrapes a value less than the previous scraped value (maybe stats is overwritten by an old haproxy process?!) and the `delta` is negative and you'll see big spikes in the graph!

we've tested OKD 4.5 router and this issue is still exist!

To resolve it, we had to parse haproxy logs using telegraf and its syslog input plugin and basicstats/historgam aggregation plugin and expose them via its promethus output plugin! A hacky workaround, but it works!

I don't know how it can be resolved! there's a Lua-based workaround here for merging the stats from multiple haproxy processes, but we haven't tested it:
https://blog.sleeplessbeastie.eu/2020/04/27/how-to-merge-multi-process-haproxy-statistics-using-lua/

**Comment by alibo** [comment]
^ Also, the response time metric of haproxy is useless!

```
 60. rtime [..BS]: the average response time in ms over the 1024 last requests
     (0 for TCP)
```

https://cbonte.github.io/haproxy-dconv/2.0/management.html#9.1

It doesn't consider all requests + you'll see a straight line for routes with low traffic!
